WalletWasabi.Tests/UnitTests/EventSourcing/InMemoryEventRepositoryTests.cs

```diff
@@ -0,0 +1,452 @@
+				await Task.WhenAll(task1, task2);
+			}
+			void WaitForConflict()
+			{
+				Assert.True(TestEventRepository.ConflictedSemaphore.Wait(_semaphoreWaitTimeout));
+			}
+			TestEventRepository.AppendedCallback = conflictAfter switch
+			{
+				nameof(TestInMemoryEventRepository.AppendedSemaphore) => WaitForConflict,
+				_ => throw new ApplicationException($"unexpected value conflictAfter: '{conflictAfter}'"),
+			};
+
+			// Assert
+			await Assert.ThrowsAsync<OptimisticConcurrencyException>(AppendInParallel);
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+						.Cast<TestWrappedEvent>().SequenceEqual(events_1));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Theory]
+		[InlineData(nameof(TestInMemoryEventRepository.AppendedSemaphore))]
+		public async Task AppendEvents_CriticalAppendConflicts_Async(string conflictAfter)
+		{
+			// Arrange
+			var events_1 = new[] { new TestWrappedEvent(1, "a"), new TestWrappedEvent(2, "a") };
+			var events_2 = new[] { new TestWrappedEvent(3, "b"), new TestWrappedEvent(4, "b") };
+
+			// Act
+			async Task Append1()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_1!);
+			}
+			async Task Append2()
+			{
+				switch (conflictAfter)
+				{
+					case nameof(TestInMemoryEventRepository.AppendedSemaphore):
+						Assert.True(TestEventRepository.AppendedSemaphore.Wait(_semaphoreWaitTimeout));
+						break;
+
+					default:
+						throw new ApplicationException($"unexpected value conflictAfter: '{conflictAfter}'");
+				}
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_2!);
+			}
+			async Task AppendInParallel()
+			{
+				var task1 = Task.Run(Append1);
+				var task2 = Task.Run(Append2);
+				await Task.WhenAll(task1, task2);
+			}
+			void WaitForNoConflict()
+			{
+				Assert.False(TestEventRepository.ConflictedSemaphore.Wait(_semaphoreWaitTimeout));
+			}
+			TestEventRepository.AppendedCallback = conflictAfter switch
+			{
+				nameof(TestInMemoryEventRepository.AppendedSemaphore) => WaitForNoConflict,
+				_ => throw new ApplicationException($"unexpected value conflictAfter: '{conflictAfter}'"),
+			};
+
+			// no conflict
+			await AppendInParallel();
+
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+						.Cast<TestWrappedEvent>().SequenceEqual(events_1.Concat(events_2)));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Theory]
+		[InlineData(nameof(TestInMemoryEventRepository.ValidatedCallback))]
+		[InlineData(nameof(TestInMemoryEventRepository.AppendedCallback))]
+		public async Task ListEventsAsync_ConflictWithAppending_Async(string listOnCallback)
+		{
+			// Arrange
+			var events_1 = new[] { new TestWrappedEvent(1, "a"), new TestWrappedEvent(2, "a") };
```

Comment:
nit: I don't think we use underscores too much:

```suggestion
			var events1 = new[] { new TestWrappedEvent(1, "a"), new TestWrappedEvent(2, "a") };
```

Reply:
removed